### ISSUE TITLE
Blazor WebAssembly App - API call from browser - Sorry, there's nothing at this address

### ISSUE THREAD

**Issue by Ogglas** [dotnet/aspnetcore#32522]
### Describe the bug

I created a new Blazor WebAssembly App with Microsoft Visual Studio 2019 Version 16.9.4 with Target Framework .NET 5.0 and ASP.NET Core Hosted.

The problem is that if I make an API call from the browser when the site is published to an Azure Web App I get an error response like this:

> Sorry, there's nothing at this address.

[![enter image description here][1]][1]

If I make the request with "Empty cache and hard reload" everything works as expected.

[![enter image description here][2]][2]

The request always work on `localhost`.

It is not limited to images loaded via API, same result for `JSON response`. 

[![enter image description here][3]][3]

[![enter image description here][4]][4]

How can I tell `Blazor` to not use/ignore routing for URLs that contains `/api/`?

I have read about `ASP.NET Core Blazor routing` but not found anything there.

https://docs.microsoft.com/en-us/aspnet/core/blazor/fundamentals/routing?view=aspnetcore-5.0

The error message is from the `App.razor` file:

    <CascadingAuthenticationState>
        <Router AppAssembly="@typeof(Program).Assembly" PreferExactMatches="@true">
            <Found Context="routeData">
                <RouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)" />
            </Found>
            <NotFound>
                <LayoutView Layout="@typeof(MainLayout)">
                    <p>Sorry, there's nothing at this address.</p>
                </LayoutView>
            </NotFound>
        </Router>
    </CascadingAuthenticationState>
    <MatPortalHost></MatPortalHost>

Same result if I use default value in `App.razor`:

    <Found Context="routeData">
        <AuthorizeRouteView RouteData="@routeData" DefaultLayout="@typeof(MainLayout)">
            <NotAuthorized>
                @if (!context.User.Identity.IsAuthenticated)
                {
                    <RedirectToLogin />
                }
                else
                {
                    <p>You are not authorized to access this resource.</p>
                }
            </NotAuthorized>
        </AuthorizeRouteView>
    </Found>

ImagesController:

    [Route("api/[controller]")]
    [ApiController]
    [Authorize]
    public class ImagesController : ControllerBase
    {
        private readonly ApplicationDbContext _context;
    
        public ImagesController(ApplicationDbContext context)
        {
            _context = context;
        }
    
        [HttpGet("{id}/file")]
        [AllowAnonymous]
        public async Task<ActionResult> GetImageDataFile(int id)
        {
            var image = await _context.Images.FindAsync(id);
    
            if (image == null)
            {
                return NotFound();
            }
    
            return File(image.Data, image.ContentType);
        }

`Program.cs` for `Blazor.Client`:

    namespace Blazor.Client
    {
        public class Program
        {
            public static async Task Main(string[] args)
            {
                var builder = WebAssemblyHostBuilder.CreateDefault(args);
                builder.RootComponents.Add<App>("#app");
    
                builder.Services.AddHttpClient("Blazor.ServerAPI", client => client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                    .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();
    
                builder.Services.AddHttpClient<ImagesHttpClient>(client =>
                    client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress))
                    .AddHttpMessageHandler<BaseAddressAuthorizationMessageHandler>();
    
                builder.Services.AddHttpClient<PostsAnonymousHttpClient>(client =>
                    client.BaseAddress = new Uri(builder.HostEnvironment.BaseAddress));
    
                // Supply HttpClient instances that include access tokens when making requests to the server project
                builder.Services.AddScoped(sp => sp.GetRequiredService<IHttpClientFactory>().CreateClient("Blazor.ServerAPI"));
    
                builder.Services.AddApiAuthorization();
    
                builder.Services.AddMatBlazor();
    
                await builder.Build().RunAsync();
            }
        }
    }

`Startup.cs` for `Blazor.Server`, not modified at all:

    namespace Blazor.Server
    {
        public class Startup
        {
            public Startup(IConfiguration configuration)
            {
                Configuration = configuration;
            }
    
            public IConfiguration Configuration { get; }
    
            // This method gets called by the runtime. Use this method to add services to the container.
            // For more information on how to configure your application, visit https://go.microsoft.com/fwlink/?LinkID=398940
            public void ConfigureServices(IServiceCollection services)
            {
                services.AddDbContext<ApplicationDbContext>(options =>
                    options.UseSqlServer(
                        Configuration.GetConnectionString("DefaultConnection")));
    
                services.AddDatabaseDeveloperPageExceptionFilter();
    
                services.AddDefaultIdentity<ApplicationUser>(options => options.SignIn.RequireConfirmedAccount = true)
                    .AddEntityFrameworkStores<ApplicationDbContext>();
    
                services.AddIdentityServer()
                    .AddApiAuthorization<ApplicationUser, ApplicationDbContext>();
    
                services.AddAuthentication()
                    .AddIdentityServerJwt();
    
                services.AddControllersWithViews();
                services.AddRazorPages();
            }
    
            // This method gets called by the runtime. Use this method to configure the HTTP request pipeline.
            public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
            {
                if (env.IsDevelopment())
                {
                    app.UseDeveloperExceptionPage();
                    app.UseMigrationsEndPoint();
                    app.UseWebAssemblyDebugging();
                }
                else
                {
                    app.UseExceptionHandler("/Error");
                    // The default HSTS value is 30 days. You may want to change this for production scenarios, see https://aka.ms/aspnetcore-hsts.
                    app.UseHsts();
                }
    
                app.UseHttpsRedirection();
                app.UseBlazorFrameworkFiles();
                app.UseStaticFiles();
    
                app.UseRouting();
    
                app.UseIdentityServer();
                app.UseAuthentication();
                app.UseAuthorization();
    
                app.UseEndpoints(endpoints =>
                {
                    endpoints.MapRazorPages();
                    endpoints.MapControllers();
                    endpoints.MapFallbackToFile("index.html");
                });
            }
        }
    }

Publish config looks like this:

[![enter image description here][5]][5]

  [1]: https://i.stack.imgur.com/cUgxm.png
  [2]: https://i.stack.imgur.com/GhMH0.png
  [3]: https://i.stack.imgur.com/CF3vL.png
  [4]: https://i.stack.imgur.com/iIjbB.png
  [5]: https://i.stack.imgur.com/owNv3.png

### To Reproduce

Create a new Blazor WebAssembly App using Microsoft Visual Studio 2019 Version 16.9.4 with Target Framework .NET 5.0 and ASP.NET Core Hosted. Add a new Controller with a GET request with json or file response and publish it to a Azure App Service. 

### Further technical details
- ASP.NET Core version - .NET 5.0
- Include the output of `dotnet --info` - Se below
- The IDE (VS / VS Code/ VS4Mac) you're running on, and its version -  Microsoft Visual Studio 2019 Version 16.9.4

.NET SDK (reflecting any global.json):
 Version:   5.0.202
 Commit:    db7cc87d51

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19041
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\5.0.202\

Host (useful for support):
  Version: 5.0.5
  Commit:  2f740adc14

.NET SDKs installed:
  3.1.301 [C:\Program Files\dotnet\sdk]
  5.0.102 [C:\Program Files\dotnet\sdk]
  5.0.202 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 2.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 5.0.5 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

To install additional .NET runtimes or SDKs:
  https://aka.ms/dotnet-download

**Comment by javiercn** [comment]
@Ogglas thanks for contacting us.

There's a lot of code for us to grok here and unfortunately its very easy to miss something missing. You could register the fallback route as follows:
`MapFallbackToFile("{*path:regex((?!^api/.*)}","Index.html")` to see if it fixes your issue (disclaimer, you'll have to play with the regex to make sure it works in all your scenarios)

**Comment by Ogglas** [comment]
@javiercn thanks for your reply!

Yes there is a lot of code but I wanted to give you as much information as possible.

I will try to sort out the details below:

Your current regex gives an exception due to a missing `)`

> An unhandled exception occurred while processing the request.
> RegexParseException: Invalid pattern '(?!^api/.*' at offset 10. Not enough )'s.

If I fix this all I get is this when loading the page:

> This localhost page can’t be found No webpage was found for the web address: https://localhost:12345/

Tried writing my own regex that works on regex101 but when running the application I do not get the desired result.

https://regex101.com/r/qmftyc/1

Based on your answer I found the following threads:

https://github.com/dotnet/aspnetcore/issues/24967

https://stackoverflow.com/q/61268117/3850405

I implemented the following code in `Startup.cs` for `Blazor.Server`:

```
app.UseEndpoints(endpoints =>
{
    endpoints.MapRazorPages();
    endpoints.MapControllers();
    endpoints.Map("api/{**slug}", HandleApiFallback);
    endpoints.MapFallbackToFile("index.html");
});

private Task HandleApiFallback(HttpContext context)
{
    context.Response.StatusCode = StatusCodes.Status404NotFound;
    return Task.CompletedTask;
}
```

This works really well on localhost:

![image](https://user-images.githubusercontent.com/4015237/117714883-da17a600-b1d7-11eb-83ee-e39dfea82197.png)

However If I deploy this code I will get the same error as before.

![image](https://user-images.githubusercontent.com/4015237/117714263-f5ce7c80-b1d6-11eb-881b-7be2db4bc28b.png)

If I use "Empty cache and hard reload" I will get the expected result.

![image](https://user-images.githubusercontent.com/4015237/117714333-10a0f100-b1d7-11eb-97ae-24b50e48a0e4.png)

If I then load the domain and then try to load the same URL I will again receive the same error` Sorry, there's nothing at this address.`

Summary:

Here is the blotted out URL: https://ogglas.com/api/images/1/file. If you visit it without first loading the domain it will work. If you first load https://ogglas.com then you will see the error above. Tried with different browsers, incognito and different devices and the result is the same. Use Empty cache and hard reload and it will work again.

I have also written a question about this on SO but so far I can't understand why this is happening.

https://stackoverflow.com/q/67452439/3850405

**Comment by Ogglas** [comment]
@javiercn  This was a real pain to debug but I eventually found the problem.

**TLDR**

I think you have a bug when using `Progressive Web Application`. Removed the line below from `Blazor\Client\wwwroot\index.html` and it works:

    <script>navigator.serviceWorker.register('service-worker.js');</script>

**Original:**

To have some sort of start I had to reproduce the error locally. I did this by publishing my app to my local IIS instead of IIS Express. There were a few problems with certs and `LocalDB` but I was able to reproduce the error locally after a while.

https://stackoverflow.com/q/67495195/3850405

I then tried to create a new application to see if the error existed there. In a new application everything worked as expected. I then tried to step back in `Git` to see when the error was introduced. I eventually ended up at the `Initial commit` and the error was still there.

To find the actual error I first tried with `WinMerge` to compare the folders but in order for that to work I would have to name the projects identically since every `namespace` would show a diff otherwise. 

[![enter image description here][1]][1]

Luckily there is Meld that can ignore specific words when comparing files and folders. I added two Text Filters under Meld Preferences that matched the namespaces I wanted to ignore and then compared the result. 

https://stackoverflow.com/a/46162831/3850405

Now nearly everything was identical except from the `Client\wwwroot\` folder. Looking at `index.html` I noticed a big difference:

[![enter image description here][2]][2]

When creating the app I must have checked `Progressive Web Application` due to the `serviceWorker` above. 

[![enter image description here][3]][3]

When I removed the line everything started working as expected.

    <script>navigator.serviceWorker.register('service-worker.js');</script>

I then tried to create a new project with `Progressive Web Application` set and deployed it to my local IIS with these settings:

[![enter image description here][4]][4]

After loading the site I could not reach `https://localhost/_framework/blazor.boot.json` without seeing `Sorry, there's nothing at this address.`. 

  [1]: https://i.stack.imgur.com/8VzQB.png
  [2]: https://i.stack.imgur.com/cD68y.png
  [3]: https://i.stack.imgur.com/ctzSN.png
  [4]: https://i.stack.imgur.com/Wdzwi.png

**Comment by SteveSandersonMS** [comment]
If you're trying to support navigation requests inside a PWA, you have to update the config in your `service-worker.published.js` file to include the additional endpoints. See https://github.com/dotnet/aspnetcore/blob/main/src/ProjectTemplates/Web.ProjectTemplates/content/ComponentsWebAssembly-CSharp/Client/wwwroot/service-worker.published.js#L47-L49 - you should update that logic so that it won't serve `index.html` for the server-rendered endpoints.

**Comment by Ogglas** [comment]
@SteveSandersonMS Thanks, this will surely help! I noticed that only `/connect/` and `/Identity/` were present by default. Shouldn't `/authentication/` be present as well for `login-callback` and `logout-callback`?

**Comment by javiercn** [comment]
@Ogglas `/authentication/login-callback` and `logout-callback` are client-side navigations.

**Comment by Ogglas** [comment]
@javiercn Thanks, closing the issue now then.
